modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ManagedLocalResult.java

```
@@ -43,71 +86,546 @@
+        this.expressions = expressions;
+        this.memTracker = memTracker;
 
-        this.mem = memTracker;
+        if (ses != null && ses.getQueryContext() != null) {
+            QueryContext qctx = (QueryContext)ses.getQueryContext();
+
+            ctx = qctx.context();
+
+            assert memTracker == null || ctx != null; // Context should be set when we track memory.
+        }
     }
 
-    /** {@inheritDoc} */
-    @Override protected void onUpdate(ValueRow distinctRowKey, Value[] oldRow, Value[] row) {
+    /** */
+    private boolean isEnoughMemoryAfterReservation(ValueRow distinctRowKey, Value[] oldRow, Value[] row) {
         assert !isClosed();
         assert row != null;
 
-        long memory;
+        if (memTracker == null)
+            return true; // No memory management set.
+
+        long memory = calculateMemoryDelta(distinctRowKey, oldRow, row);
+
+        memReserved += memory;
+
+        if (memory < 0) {
+            memTracker.released(-memory);
+
+            return true;
+        }
+        else
+            return memTracker.reserved(memory);
+    }
+
+    @Override public boolean isLazy() {
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void setMaxMemoryRows(int maxValue) {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public H2ManagedLocalResult createShallowCopy(SessionInterface targetSession) {
+        if (containsLobs) {
+            return null;
+        }
+
+        ResultExternal e2 = null;
+
+        if (external != null) {
+            e2 = external.createShallowCopy();
+
+            if (e2 == null)
+                return null;
+        }
+
+        H2ManagedLocalResult cp = new H2ManagedLocalResult();
+
+        cp.session = (Session)targetSession;
+        cp.visibleColumnCount = this.visibleColumnCount;
+        cp.expressions = this.expressions;
+        cp.rowId = -1;
+        cp.rowCount = this.rowCount;
+        cp.rows = this.rows;
```

Really? rows can leak to another Session?

Is it always same session and we can safely copy memoryTracker? Or this is never called and we can throw UnsupportedOperationException?

---

I expect this method is called only when we cache subquery result (see i.e. `ConditionInSelect`) within the single query. So, it is safe to copy all these fields.